Bug fixes and a change to license request logic

Mbucari opened the pull request:
## [Fix dash files not being saved](https://github.com/rmcrackan/Libation/commit/28ba62aead001dd0245976a0552eba8f57351f1b) (#1236)

I fixed this for aaxc files, but then broke it for dash files. It now works for all encrypted file types.

## [Pad final chapter to prevent truncation from incorrect chapter info](https://github.com/rmcrackan/Libation/commit/b11a4887d7ae919428ce7ea522440c1b444dbd7e) (#1246)

As I described in that issue, incorrect chapter markers which are shorter than the audiobook will no longer truncate the audio file.

## [Use Libation settings to decide which DRM is downloaded.](https://github.com/rmcrackan/Libation/commit/19db226f5ac778e46cf65ecc5f36820d27badff8)

Since I added the fine-grained options for downloading widevine content in 12.4, it makes sense to only use those options to dictate which content Libation tries to download. I removed the fallback to traditional ADRM. If for whatever reason widevine content becomes unavailable, users can uncheck the "Use widevine DRM" option to revert back to the ADRM-onyl downloads.

rmcrackan commented on Source/LibationUiBase/LibationContributor.cs:
```diff
@@ -41,7 +41,8 @@ public static IEnumerable<LibationContributor> AdditionalContributors
 			GitHubUser("patienttruth"),
-			GitHubUser("stickystyle")
+			GitHubUser("stickystyle"),
+			GitHubUser("cherez"),
```
good memory

Mbucari commented:
## [Fix occasional error of audio downloads hanging.](https://github.com/rmcrackan/Libation/pull/1250/commits/10c01f4147f49b5351862b5fe153af322deea89d)

I think I discovered the source of the bug which was causing some people's prowess to hang.
First of all, the step runner would hang with an unhandled exception. I fixed that.
Secondly, sometimes Audible would return more than 1 key, and only the second key was valid for the file. We now pass all possible keys down to AaxDecrypter and let it figure out which one to use.

## [Change ApiExtended to always allow provide login option](https://github.com/rmcrackan/Libation/pull/1250/commits/f35c82d59dfb63043f3711b6617ab75bcb69245d)

Previously, only some calls to `ApiExtended.CreateAsync()` would prompt users to login if necessary. Other calls would only work if the account already had a valid identity, and they would throw exceptions otherwise.

Changed ApiExtended so that the UI registers a static ILoginChoiceEager factory delegate that ApiExtended will use in the event that a login is required.

## [Ask users to clear the accounts when enabling widevine](https://github.com/rmcrackan/Libation/pull/1250/commits/ef67ae9d6a292fe5bd69a7df588888751d9bc162)

First, changed the "Use widevine DRM" default setting to "false". This should be opt-in.

Second, when a user does opt in and they still have iphone accounts, Libation asks them for permission to clear their accounts and prompt them for a login on the next download. If the choose "yes", Libation deletes and re-adds their account. Because of the changes made above, any account api action will trigger the login prompt.

rmcrackan commented:
Nice. This was quite a bit in a short amount of time.

> I discovered another bug that really needs to be fixed, but I'm done for a while.

Want me to release now? Or wait for that fix?

Mbucari commented:
> > I discovered another bug that really needs to be fixed, but I'm done for a while.
> 
> Want me to release now? Or wait for that fix?

Sorry for the confusion. I was referring to the bug that I fixed in this PR. I was just explaining why I converted to draft and added more.

> Nice. This was quite a bit in a short amount of time.

Yeah, it's actually a pretty critical bug which is why I was determined to fix it ASAP. It was just dumb luck that I didn't realize what was going on sooner, but this is probably affecting a lot of people.

Mbucari commented on Source/AudibleUtilities/ApiExtended.cs:
```diff
@@ -24,52 +26,46 @@ public class ApiExtended
-				LocaleName = localeName,
-			});
+				if (LoginChoiceFactory is null)
+					throw new InvalidOperationException($"The UI module must first set {LoginChoiceFactory} before attempting to create the api");
```
Oops, that should be a nameof() call. Would you mind fixing that please?

rmcrackan commented:
Something fucky is going on with my browser. I'm trying to add a comment to DownloadOptions.Factory.cs ~line 82 : Did you want to move this large explanation elsewhere?

rmcrackan commented:
I notified/closed these. Lemme know if this PR resolved any other tix: #1249 , #1246 , #1236

Mbucari commented:
> Something fucky is going on with my browser. I'm trying to add a comment to DownloadOptions.Factory.cs ~line 82 : Did you want to move this large explanation elsewhere?

The explanation is no longer relevant. Which license gets used is now determined by libation settings, no longer by heuristics.

Mbucari commented:
@rmcrackan this will improve #1228.

Also, the old fix for #1228 would often download the older, non widevine audio files if they were roughly the same size as the widevine files. Now, when "Use widevine DRM" is checked, Libation always tries to download widevine content. If it fails, it logs the error and will not fall back to the old ADRM license request.
